package/MDAnalysis/core/groups.py

```diff
@@ -640,16 +648,57 @@ def center(self, weights, pbc=None):
         else:
             xyz = atoms.positions
 
-        return np.average(xyz, weights=weights, axis=0)
-
-    def center_of_geometry(self, pbc=None):
+        if compound.lower() == 'group':
+            return np.average(xyz, weights=weights, axis=0)
+        elif compound.lower() == 'residues':
+            compound_indices = atoms.resindices
+            n_compounds = atoms.n_residues
+        elif compound.lower() == 'segments':
+            compound_indices = atoms.segindices
+            n_compounds = atoms.n_segments
+        else:
+            raise ValueError("Unrecognized compound definition: {}\nPlease use"
+                             " one of 'group', 'residues', or 'segments'."
+                             "".format(compound))
+
+        # Sort positions and masses by compound index:
+        sort_indices = np.argsort(compound_indices)
+        compound_indices = compound_indices[sort_indices]
+        positions = atoms.positions[sort_indices]
+        weights = weights[sort_indices]
+        # Allocate output array:
+        coms = np.zeros((n_compounds, 3), dtype=np.float32)
+        # Get sizes of compounds:
+        unique_compound_indices, compound_sizes = np.unique(compound_indices,
+                                                            return_counts=True)
+        unique_compound_sizes = np.unique(compound_sizes)
+        # Compute centers per compound for each compound size:
+        for compound_size in unique_compound_sizes:
+            compound_mask = compound_sizes == compound_size
+            _compound_indices = unique_compound_indices[compound_mask]
+            atoms_mask = np.isin(compound_indices, _compound_indices)
+            _positions = positions[atoms_mask].reshape((-1, compound_size, 3))
+            _weights = weights[atoms_mask].reshape((-1, compound_size))
+            _coms = (_positions * _weights[:, :, None]).sum(axis=1)
+            _coms /= _weights.sum(axis=1)[:, None]
+            coms[compound_mask] = _coms
+        return coms
```

Comment:
PBC are not handled for residues or segments!

Reply:
Yes they are currently not handled correctly. I wrongly assumed that just packing all atoms into the box would work. But yeah this is wrong

Reply:
This means we should also test pbc conditions in the unit tests